pkg/crypto/certificatemanager/certificate_manager.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package certificatemanager
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+
+	"github.com/cilium/cilium/pkg/policy/api"
+)
+
+const (
+	caDefaultName      = "ca.crt"
+	publicDefaultName  = "tls.crt"
+	privateDefaultName = "tls.key"
+)
+
+type k8sClient interface {
+	GetSecrets(ctx context.Context, ns, name string) (map[string][]byte, error)
+}
+
+// Manager will manage the way certificates are retrieved based in the given
+// k8sClient and rootPath.
+type Manager struct {
+	rootPath  string
+	k8sClient k8sClient
+}
+
+// NewManager returns a new manager.
+func NewManager(certsRootPath string, k8sClient k8sClient) *Manager {
+	return &Manager{
+		rootPath:  certsRootPath,
+		k8sClient: k8sClient,
+	}
+}
+
+// GetTLSContext returns a new ca, public and private certificates found based
+// in the given api.TLSContext.
+func (m *Manager) GetTLSContext(ctx context.Context, tlsCtx *api.TLSContext, ns string) (ca, public, private string, err error) {
+	var caBytes, publicBytes, privateBytes []byte
+
+	caMustExist := false
+	caName := caDefaultName
+	if tlsCtx.TrustedCA != "" {
+		caName = tlsCtx.TrustedCA
+		caMustExist = true
+	}
+
+	publicMustExist := false
+	publicName := publicDefaultName
+	if tlsCtx.Certificate != "" {
+		publicName = tlsCtx.Certificate
+		publicMustExist = true
+	}
+
+	privateMustExist := false
+	privateName := privateDefaultName
+	if tlsCtx.PrivateKey != "" {
+		privateName = tlsCtx.PrivateKey
+		privateMustExist = true
+	}
+
+	if tlsCtx.Secret != nil {
+		if tlsCtx.Secret.Namespace != "" {
+			ns = tlsCtx.Secret.Namespace
+		}
+		if tlsCtx.Secret.Name == "" {
+			err = fmt.Errorf("Missing Secret name")
+			return "", "", "", err
+		}
+		name := tlsCtx.Secret.Name
+
+		// Give priority to local certificates
+		certPath := filepath.Join(m.rootPath, ns, name)
+		files, ioErr := ioutil.ReadDir(certPath)
+		if ioErr != nil {
+			err = fmt.Errorf("Certificates directory %s not found (%s)", certPath, ioErr)
```

Comment:
nit:
```suggestion
			err = fmt.Errorf("cannot read certificates directory %s  (%s)", certPath, ioErr)
```

(If it's really `not found` then the ioErr should say this. Or permission denied, or whatever the error really is.)

Reply:
Fixed.